File: packages/mobile-fetch/ios/HybridStreamFetchBuilder.swift
import Foundation
import NitroModules

class HybridStreamFetchBuilder: HybridStreamFetchBuilderSpec {
  private let url: String
  private let session: URLSession
  private var request: URLRequest
  private var onResponseCallback: ((_ info: StreamFetchResponseInfo) -> Void)?
  private var onChunkCallback: ((_ bytes: ArrayBuffer) -> Void)?
  private var onCompleteCallback: (() -> Void)?
  private var onErrorCallback: ((_ message: String) -> Void)?

  init(url: String, session: URLSession) {
    self.url = url
    self.session = session
    self.request = URLRequest(url: URL(string: url) ?? URL(fileURLWithPath: "/"))
    self.request.cachePolicy = .reloadIgnoringLocalCacheData
    super.init()
  }

  func setMethod(httpMethod: String) throws {
    request.httpMethod = httpMethod
  }

  func addHeader(name: String, value: String) throws {
    request.addValue(value, forHTTPHeaderField: name)
  }

  func setBodyString(body: String) throws {
    request.httpBody = body.data(using: .utf8)
  }

  func setBodyBytes(body: ArrayBuffer) throws {
    request.httpBody = body.toData(copyIfNeeded: true)
  }

  func onResponse(callback: @escaping (_ info: StreamFetchResponseInfo) -> Void) throws {
    onResponseCallback = callback
  }

  func onChunk(callback: @escaping (_ bytes: ArrayBuffer) -> Void) throws {
    onChunkCallback = callback
  }

  func onComplete(callback: @escaping () -> Void) throws {
    onCompleteCallback = callback
  }

  func onError(callback: @escaping (_ message: String) -> Void) throws {
    onErrorCallback = callback
  }

  func build() throws -> any HybridStreamFetchRequestSpec {
    guard URL(string: url) != nil else {
      throw RuntimeError.error(withMessage: "Invalid URL")
    }
    let delegate = StreamFetchDelegate(
      onResponse: onResponseCallback,
      onChunk: onChunkCallback,
      onComplete: onCompleteCallback,
      onError: onErrorCallback
    )
    let task = session.dataTask(with: request)
    task.delegate = delegate
    return HybridStreamFetchRequest(task: task, delegate: delegate)
  }
}

private final class StreamFetchDelegate: NSObject, URLSessionDataDelegate {
  private let onResponse: ((_ info: StreamFetchResponseInfo) -> Void)?
  private let onChunk: ((_ bytes: ArrayBuffer) -> Void)?
  private let onComplete: (() -> Void)?
  private let onError: ((_ message: String) -> Void)?
  private var httpResponse: HTTPURLResponse?
  private var finished = false

  init(
    onResponse: ((_ info: StreamFetchResponseInfo) -> Void)?,
    onChunk: ((_ bytes: ArrayBuffer) -> Void)?,
    onComplete: (() -> Void)?,
    onError: ((_ message: String) -> Void)?
  ) {
    self.onResponse = onResponse
    self.onChunk = onChunk
    self.onComplete = onComplete
    self.onError = onError
  }

  func urlSession(
    _ session: URLSession,
    dataTask: URLSessionDataTask,
    didReceive response: URLResponse,
    completionHandler: @escaping (URLSession.ResponseDisposition) -> Void
  ) {
    guard let httpResponse = response as? HTTPURLResponse else {
      completionHandler(.cancel)
      return
    }
    self.httpResponse = httpResponse
    let headers = httpResponse.allHeaderFields.compactMap { key, value -> StreamFetchHeader? in
      guard let name = key as? String else { return nil }
      return StreamFetchHeader(key: name, value: String(describing: value))
    }
    onResponse?(
      StreamFetchResponseInfo(
        url: httpResponse.url?.absoluteString ?? "",
        status: Double(httpResponse.statusCode),
        statusText: HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode),
        headers: headers
      )
    )
    completionHandler(.allow)
  }

  func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
    guard !data.isEmpty else { return }
    do {
      let buffer = try ArrayBuffer.copy(data: data)
      onChunk?(buffer)
    } catch {
      finishError(error.localizedDescription)
    }
  }

  func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
    if let error {
      finishError(error.localizedDescription)
      return
    }
    finishComplete()
  }

  private func finishComplete() {
    guard !finished else { return }
    finished = true
    onComplete?()
  }

  private func finishError(_ message: String) {
    guard !finished else { return }
    finished = true
    onError?(message)
  }
}

File: packages/mobile-fetch/android/src/main/java/com/margelo/nitro/mobilefetch/HybridStreamFetchBuilder.kt
package com.margelo.nitro.mobilefetch

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import okhttp3.Headers
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

@DoNotStrip
@Keep
class HybridStreamFetchBuilder(
  private val url: String,
  private val client: OkHttpClient,
) : HybridStreamFetchBuilderSpec() {
  private var method = "GET"
  private val headers = Headers.Builder()
  private var bodyBytes: ByteArray? = null
  private var onResponseCallback: ((info: StreamFetchResponseInfo) -> Unit)? = null
  private var onChunkCallback: ((bytes: ArrayBuffer) -> Unit)? = null
  private var onCompleteCallback: (() -> Unit)? = null
  private var onErrorCallback: ((message: String) -> Unit)? = null

  override fun setMethod(httpMethod: String) {
    this.method = httpMethod.uppercase()
  }

  override fun addHeader(name: String, value: String) {
    headers.add(name, value)
  }

  override fun setBodyString(body: String) {
    bodyBytes = body.toByteArray(Charsets.UTF_8)
  }

  override fun setBodyBytes(body: ArrayBuffer) {
    bodyBytes = body.toByteArray()
  }

  override fun onResponse(callback: (info: StreamFetchResponseInfo) -> Unit) {
    onResponseCallback = callback
  }

  override fun onChunk(callback: (bytes: ArrayBuffer) -> Unit) {
    onChunkCallback = callback
  }

  override fun onComplete(callback: () -> Unit) {
    onCompleteCallback = callback
  }

  override fun onError(callback: (message: String) -> Unit) {
    onErrorCallback = callback
  }

  override fun build(): HybridStreamFetchRequestSpec {
    val requestHeaders = headers.build()
    val body =
      bodyBytes?.let { bytes ->
        val mediaType = requestHeaders["Content-Type"]?.toMediaTypeOrNull()
        bytes.toRequestBody(mediaType)
      }
    val request =
      Request.Builder()
        .url(url)
        .headers(requestHeaders)
        .method(method, if (methodAllowsBody(method)) body ?: ByteArray(0).toRequestBody(null) else null)
        .build()
    return HybridStreamFetchRequest(
      client = client,
      request = request,
      onResponse = onResponseCallback,
      onChunk = onChunkCallback,
      onComplete = onCompleteCallback,
      onError = onErrorCallback,
    )
  }

  private fun methodAllowsBody(method: String): Boolean {
    return method != "GET" && method != "HEAD"
  }
}
